# Projeto #2 - Classificador supervisionado

Antes de começar, leia as [Instruções](https://github.com/thvmm/pos-ds-ia/tree/master/projeto_2#instru%C3%A7%C3%B5es) e os [Critérios de Avaliação](https://github.com/thvmm/pos-ds-ia/tree/master/projeto_2#crit%C3%A9rios-de-avalia%C3%A7%C3%A3o)


### 1) Qual a base escolhida?

In [322]:
Forest covertypes

SyntaxError: invalid syntax (<ipython-input-322-fcb1a410af2a>, line 1)

### 2) **(10%)** Pré-processamento: entendimento do conjunto de dados
- Quais são minhas features?
- Quais são minhas classes?
- Como estão distribuidas minhas classes?
- Checagem se os valores estão dentro de um limite permitido ou razoável.
- Tratamento de valores ausentes por eliminação ou substituição.
- Conversão do tipo de dados.

In [ ]:
import pandas as pd  
import numpy as np  
import matplotlib.pyplot as plt  
import seaborn as seabornInstance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics
from sklearn import tree
from sklearn import preprocessing
import time
import sklearn
from collections import Counter
from sklearn.model_selection import cross_val_score
%matplotlib inline

In [ ]:
data = pd.read_table(r'C:\Users\braga\Google Drive\Trabalho2\covtype.data',sep=',',header=None)

In [323]:
data.columns = ['Elevation','Aspect','Slope','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Hillshade_9am','Hillshade_Noon','Hillshade_3pm','Horizontal_Distance_To_Fire_Points','Wilderness_Area_1','Wilderness_Area_2','Wilderness_Area_3','Wilderness_Area_4','Soil_Type_1','Soil_Type_2','Soil_Type_3','Soil_Type_4','Soil_Type_5','Soil_Type_6','Soil_Type_7','Soil_Type_8','Soil_Type_9','Soil_Type_10','Soil_Type_11','Soil_Type_12','Soil_Type_13','Soil_Type_14','Soil_Type_15','Soil_Type_16','Soil_Type_17','Soil_Type_18','Soil_Type_19','Soil_Type_20','Soil_Type_21','Soil_Type_22','Soil_Type_23','Soil_Type_24','Soil_Type_25','Soil_Type_26','Soil_Type_27','Soil_Type_28','Soil_Type_29','Soil_Type_30','Soil_Type_31','Soil_Type_32','Soil_Type_33','Soil_Type_34','Soil_Type_35','Soil_Type_36','Soil_Type_37','Soil_Type_38','Soil_Type_39','Soil_Type_40','Cover_Type']
data.shape
#Temos um total de 581012 linhas com 55 colunas.

(581012, 55)

In [324]:
data.columns

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Hydrology',
       'Vertical_Distance_To_Hydrology', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area_1',
       'Wilderness_Area_2', 'Wilderness_Area_3', 'Wilderness_Area_4',
       'Soil_Type_1', 'Soil_Type_2', 'Soil_Type_3', 'Soil_Type_4',
       'Soil_Type_5', 'Soil_Type_6', 'Soil_Type_7', 'Soil_Type_8',
       'Soil_Type_9', 'Soil_Type_10', 'Soil_Type_11', 'Soil_Type_12',
       'Soil_Type_13', 'Soil_Type_14', 'Soil_Type_15', 'Soil_Type_16',
       'Soil_Type_17', 'Soil_Type_18', 'Soil_Type_19', 'Soil_Type_20',
       'Soil_Type_21', 'Soil_Type_22', 'Soil_Type_23', 'Soil_Type_24',
       'Soil_Type_25', 'Soil_Type_26', 'Soil_Type_27', 'Soil_Type_28',
       'Soil_Type_29', 'Soil_Type_30', 'Soil_Type_31', 'Soil_Type_32',
       'Soil_Type_33', 'Soil_Type_34', 'Soil_Type_35', 'Soil_Type_36',
       'Soil_Type_37', '

Dentro do Data Frame temos as colunas(features):
Elevation:                          Elevação em metros
Aspect:                             Aspecto em graus azimute
Slope:                              Inclinação em graus
Horizontal_Distance_To_Hydrology:   Distância horizontal ate a superficie de água mais próximas em metros
Vertical_Distance_To_Hydrology:     Distância vertical ate a superficie de água mais próximas em metros
Horizontal_Distance_To_Roadways:    Distância horizontal até a estrada mais próxima em metros
Hillshade_9am                       Índice Hillshade para formação de mapa topográfico via luz e sombra
Hillshade_Noon                      Índice Hillshade para formação de mapa topográfico via luz e sombra
Hillshade_3pm                       Índice Hillshade para formação de mapa topográfico via luz e sombra
*How Hillshade works:               https://desktop.arcgis.com/en/arcmap/10.3/tools/spatial-analyst-toolbox/how-hillshade-works.htm
Horizontal_Distance_To_Fire_Points: Distância horizontal até os focos de incêndio mais próximos em metros
Wilderness_Area 1 ate 4:            Tipos de áreas selvagens (binário)
Soil_Type 1 ate 40:                 Tipos de solos (binário)
Cover_Type:                         Tipo de floresta do local, vai de 1 a 7 (classes)
** Informações dos binários e classes estão abaixo

In [325]:
Wilderness_Area = {'Type':[1,2,3,4],'Name':['Rawah Wilderness Area','Neota Wilderness Area','Comanche Peak Wilderness Area','Cache la Poudre Wilderness Area']}
Wilderness_Area = pd.DataFrame(Wilderness_Area)
Wilderness_Area

Type                             Name
0     1            Rawah Wilderness Area
1     2            Neota Wilderness Area
2     3    Comanche Peak Wilderness Area
3     4  Cache la Poudre Wilderness Area

In [326]:
Soil_Type = {'Type':[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40],'Description':['Cathedral family - Rock outcrop complex, extremely stony','Vanet - Ratake families complex, very stony','Haploborolis - Rock outcrop complex, rubbly','Ratake family - Rock outcrop complex, rubbly','Vanet family - Rock outcrop complex complex, rubbly','Vanet - Wetmore families - Rock outcrop complex, stony','Gothic family','Supervisor - Limber families complex','Troutville family, very stony','Bullwark - Catamount families - Rock outcrop complex, rubbly','Bullwark - Catamount families - Rock land complex, rubbly','Legault family - Rock land complex, stony','Catamount family - Rock land - Bullwark family complex, rubbly','Pachic Argiborolis - Aquolis complex','Unspecified in the USFS Soil and ELU Survey','Cryaquolis - Cryoborolis complex','Gateview family - Cryaquolis complex','Rogert family, very stony','Typic Cryaquolis - Borohemists complex','Typic Cryaquepts - Typic Cryaquolls complex','Typic Cryaquolls - Leighcan family, till substratum complex','Leighcan family, till substratum, extremely bouldery','Leighcan family, till substratum - Typic Cryaquolls complex','Leighcan family, extremely stony','Leighcan family, warm, extremely stony','Granile - Catamount families complex, very stony','Leighcan family, warm - Rock outcrop complex, extremely stony','Leighcan family - Rock outcrop complex, extremely stony','Como - Legault families complex, extremely stony','Como family - Rock land - Legault family complex, extremely stony','Leighcan - Catamount families complex, extremely stony','Catamount family - Rock outcrop - Leighcan family complex, extremely stony','Leighcan - Catamount families - Rock outcrop complex, extremely stony','Cryorthents - Rock land complex, extremely stony','Cryumbrepts - Rock outcrop - Cryaquepts complex','Bross family - Rock land - Cryumbrepts complex, extremely stony','Rock outcrop - Cryumbrepts - Cryorthents complex, extremely stony','Leighcan - Moran families - Cryaquolls complex, extremely stony','Moran family - Cryorthents - Leighcan family complex, extremely stony','Moran family - Cryorthents - Rock land complex, extremely stony'],'Climatic_Zone_Code':[2,2,2,2,2,2,3,3,4,4,4,4,4,5,5,6,6,6,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,8,8,8,8,8,8],'Climatic_Zone_Description':['lowermontane','lowermontane','lowermontane','lowermontane','lowermontane','lowermontane','montanedry','montanedry','montane','montane','montane','montane','montane','montanedryandmontane','montanedryandmontane','montaneandsubalpine','montaneandsubalpine','montaneandsubalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','subalpine','alpine','alpine','alpine','alpine','alpine','alpine'],'Geologic_Zone_Code':[7,7,7,7,7,7,5,5,2,7,7,7,7,1,1,1,1,7,1,1,1,2,2,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7],'Geologic_Zone_Description':['igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','mixedsedimentary','mixedsedimentary','glacial','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','alluvium','alluvium','alluvium','alluvium','igneousandmetamorphic','alluvium','alluvium','alluvium','glacial','glacial','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic','igneousandmetamorphic']}
Soil_Type = pd.DataFrame(Soil_Type)
Soil_Type

Type                                        Description  \
0      1  Cathedral family - Rock outcrop complex, extre...   
1      2        Vanet - Ratake families complex, very stony   
2      3        Haploborolis - Rock outcrop complex, rubbly   
3      4       Ratake family - Rock outcrop complex, rubbly   
4      5  Vanet family - Rock outcrop complex complex, r...   
5      6  Vanet - Wetmore families - Rock outcrop comple...   
6      7                                      Gothic family   
7      8               Supervisor - Limber families complex   
8      9                      Troutville family, very stony   
9     10  Bullwark - Catamount families - Rock outcrop c...   
10    11  Bullwark - Catamount families - Rock land comp...   
11    12          Legault family - Rock land complex, stony   
12    13  Catamount family - Rock land - Bullwark family...   
13    14               Pachic Argiborolis - Aquolis complex   
14    15        Unspecified in the USFS Soil and ELU Survey   
15    16                   Cryaquolis - Cryoborolis complex   
16    17               Gateview family - Cryaquolis complex   
17    18                          Rogert family, very stony   
18    19             Typic Cryaquolis - Borohemists complex   
19    20        Typic Cryaquepts - Typic Cryaquolls complex   
20    21  Typic Cryaquolls - Leighcan family, till subst...   
21    22  Leighcan family, till substratum, extremely bo...   
22    23  Leighcan family, till substratum - Typic Cryaq...   
23    24                   Leighcan family, extremely stony   
24    25             Leighcan family, warm, extremely stony   
25    26   Granile - Catamount families complex, very stony   
26    27  Leighcan family, warm - Rock outcrop complex, ...   
27    28  Leighcan family - Rock outcrop complex, extrem...   
28    29   Como - Legault families complex, extremely stony   
29    30  Como family - Rock land - Legault family compl...   
30    31  Leighcan - Catamount families complex, extreme...   
31    32  Catamount family - Rock outcrop - Leighcan fam...   
32    33  Leighcan - Catamount families - Rock outcrop c...   
33    34   Cryorthents - Rock land complex, extremely stony   
34    35    Cryumbrepts - Rock outcrop - Cryaquepts complex   
35    36  Bross family - Rock land - Cryumbrepts complex...   
36    37  Rock outcrop - Cryumbrepts - Cryorthents compl...   
37    38  Leighcan - Moran families - Cryaquolls complex...   
38    39  Moran family - Cryorthents - Leighcan family c...   
39    40  Moran family - Cryorthents - Rock land complex...   

    Climatic_Zone_Code Climatic_Zone_Description  Geologic_Zone_Code  \
0                    2              lowermontane                   7   
1                    2              lowermontane                   7   
2                    2              lowermontane                   7   
3                    2              lowermontane                   7   
4                    2              lowermontane                   7   
5                    2              lowermontane                   7   
6                    3                montanedry                   5   
7                    3                montanedry                   5   
8                    4                   montane                   2   
9                    4                   montane                   7   
10                   4                   montane                   7   
11                   4                   montane                   7   
12                   4                   montane                   7   
13                   5      montanedryandmontane                   1   
14                   5      montanedryandmontane                   1   
15                   6       montaneandsubalpine                   1   
16                   6       montaneandsubalpine                   1   
17                   6       montaneandsubalpine                   7   
18                   7                 subalpine    

In [327]:
data.count()
#Nenhum valor nulo

Elevation                             581012
Aspect                                581012
Slope                                 581012
Horizontal_Distance_To_Hydrology      581012
Vertical_Distance_To_Hydrology        581012
Horizontal_Distance_To_Roadways       581012
Hillshade_9am                         581012
Hillshade_Noon                        581012
Hillshade_3pm                         581012
Horizontal_Distance_To_Fire_Points    581012
Wilderness_Area_1                     581012
Wilderness_Area_2                     581012
Wilderness_Area_3                     581012
Wilderness_Area_4                     581012
Soil_Type_1                           581012
Soil_Type_2                           581012
Soil_Type_3                           581012
Soil_Type_4                           581012
Soil_Type_5                           581012
Soil_Type_6                           581012
Soil_Type_7                           581012
Soil_Type_8                           581012
Soil_Type_

In [328]:
data.iloc[0:,0:10].describe()

Elevation         Aspect          Slope  \
count  581012.000000  581012.000000  581012.000000   
mean     2959.365301     155.656807      14.103704   
std       279.984734     111.913721       7.488242   
min      1859.000000       0.000000       0.000000   
25%      2809.000000      58.000000       9.000000   
50%      2996.000000     127.000000      13.000000   
75%      3163.000000     260.000000      18.000000   
max      3858.000000     360.000000      66.000000   

       Horizontal_Distance_To_Hydrology  Vertical_Distance_To_Hydrology  \
count                     581012.000000                   581012.000000   
mean                         269.428217                       46.418855   
std                          212.549356                       58.295232   
min                            0.000000                     -173.000000   
25%                          108.000000                        7.000000   
50%                          218.000000                       30.000000   
75%                          384.000000                       69.000000   
max                         1397.000000                      601.000000   

       Horizontal_Distance_To_Roadways  Hillshade_9am  Hillshade_Noon  \
count                    581012.000000  581012.000000   581012.000000   
mean                       2350.146611     212.146049      223.318716   
std                        1559.254870      26.769889       19.768697   
min                           0.000000       0.000000        0.000000   
25%                        1106.000000     198.000000      213.000000   
50%                        1997.000000     218.000000      226.000000   
75%                        3328.000000     231.000000      237.000000   
max                        7117.000000     254.000000      254.000000   

       Hillshade_3pm  Horizontal_Distance_To_Fire_Points  
count  581012.000000                       581012.000000  
mean      142.528263                         1980.291226  
std        38.274529                         1324.195210  
min         0.000000                            0.000000  
25%       119.000000                         1024.000000  
50%       143.000000                         1710.000000  
75%       168.000000                         2550.000000  
max       254.000000                         7173.000000

In [329]:
pd.unique(data.iloc[0:,10:54].values.ravel('K'))
#Nenhum valor fora do permitido

array([1, 0], dtype=int64)

In [330]:
data['Cover_Type'].unique()
#Nenhum valor fora do permitido

array([5, 2, 1, 7, 3, 6, 4], dtype=int64)

### 3) **(80%)** Nos blocos seguintes implemente seus classificadores (serão implementados 2 métodos diferentes).

#### 3.1) Qual método escolhido?

Linear Regression

#### 3.2) **(10%)** Baseline - Implemente seu classificador da forma mais simples possível para esse ser seu baseline

In [331]:
base_df_linear = pd.DataFrame.copy(data)
X = base_df_linear.iloc[0:,0:54]
y = base_df_linear.iloc[0:,54]

In [332]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [333]:
regressor = LinearRegression()  
regressor.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [334]:
y_pred = regressor.predict(X_test)

In [335]:
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323434740334


#### 3.3) **(20%)** Versão 1 - O que podemos fazer para melhorar nosso baseline? Aplique técnicas como redução de dimensionalidade, normalização ou outras. Compare os resultados.

In [336]:
base_df_linear2 = pd.DataFrame.copy(data)

In [337]:
base_df_linear2.iloc[0:,10:14]

Wilderness_Area_1  Wilderness_Area_2  Wilderness_Area_3  \
0                       1                  0                  0   
1                       1                  0                  0   
2                       1                  0                  0   
3                       1                  0                  0   
4                       1                  0                  0   
...                   ...                ...                ...   
581007                  0                  0                  1   
581008                  0                  0                  1   
581009                  0                  0                  1   
581010                  0                  0                  1   
581011                  0                  0                  1   

        Wilderness_Area_4  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
581007                  0  
581008                  0  
581009                  0  
581010                  0  
581011                  0  

[581012 rows x 4 columns]

In [338]:
Wilderness_Area = {}
Wilderness_Area = pd.DataFrame(Wilderness_Area)
i = 1
while i < 5:
    Wilderness_Area['Wilderness_Area_'+ str(i)] = base_df_linear2['Wilderness_Area_'+ str(i)]*i
    i = i + 1
Wilderness_Area['Wilderness_Area'] = Wilderness_Area.sum(axis=1)
Wilderness_Area = Wilderness_Area.drop(Wilderness_Area.loc[:, 'Wilderness_Area_1':'Wilderness_Area_4'].columns, axis = 1)

In [339]:
Wilderness_Area

Wilderness_Area
0                     1
1                     1
2                     1
3                     1
4                     1
...                 ...
581007                3
581008                3
581009                3
581010                3
581011                3

[581012 rows x 1 columns]

In [340]:
Wilderness_Area['Wilderness_Area'].unique()

array([1, 3, 4, 2], dtype=int64)

In [341]:
base_df_linear2.iloc[0:,14:54]

Soil_Type_1  Soil_Type_2  Soil_Type_3  Soil_Type_4  Soil_Type_5  \
0                 0            0            0            0            0   
1                 0            0            0            0            0   
2                 0            0            0            0            0   
3                 0            0            0            0            0   
4                 0            0            0            0            0   
...             ...          ...          ...          ...          ...   
581007            0            1            0            0            0   
581008            0            1            0            0            0   
581009            0            1            0            0            0   
581010            0            1            0            0            0   
581011            0            1            0            0            0   

        Soil_Type_6  Soil_Type_7  Soil_Type_8  Soil_Type_9  Soil_Type_10  ...  \
0                 0            0            0            0             0  ...   
1                 0            0            0            0             0  ...   
2                 0            0            0            0             0  ...   
3                 0            0            0            0             0  ...   
4                 0            0            0            0             0  ...   
...             ...          ...          ...          ...           ...  ...   
581007            0            0            0            0             0  ...   
581008            0            0            0            0             0  ...   
581009            0            0            0            0             0  ...   
581010            0            0            0            0             0  ...   
581011            0            0            0            0             0  ...   

        Soil_Type_31  Soil_Type_32  Soil_Type_33  Soil_Type_34  Soil_Type_35  \
0                  0             0             0             0             0   
1                  0             0             0             0             0   
2                  0             0             0             0             0   
3                  0             0             0             0             0   
4                  0             0             0             0             0   
...              ...           ...           ...           ...           ...   
581007             0             0             0             0             0   
581008             0             0             0             0             0   
581009             0             0             0             0             0   
581010             0             0             0             0             0   
581011             0             0             0             0             0   

        Soil_Type_36  Soil_Type_37  Soil_Type_38  Soil_Type_39  Soil_Type_40  
0                  0             0             0             0             0  
1                  0             0             0             0             0  
2                  0             0             0             0             0  
3                  0             0             0             0             0  
4                  0             0             0             0             0  
...              ...           ...           ...           ...           ...  
581007             0             0             0             0             0  
581008             0             0             0             0             0  
581009             0             0             0             0             0  
581010             0             0             0             0             0  
581011             0             0             0             0             0  

[581012 rows x 40 columns]

In [342]:
Soil_Type = {}
Soil_Type = pd.DataFrame(Soil_Type)
i = 1
while i < 41:
    Soil_Type['Soil_Type_'+ str(i)] = base_df_linear2['Soil_Type_'+ str(i)]*i
    i = i + 1
Soil_Type['Soil_Type'] = Soil_Type.sum(axis=1)
Soil_Type = Soil_Type.drop(Soil_Type.loc[:, 'Soil_Type_1':'Soil_Type_40'].columns, axis = 1)

In [343]:
Soil_Type['Soil_Type'].unique()

array([29, 12, 30, 18, 16, 20, 24, 23, 40, 19,  8, 22, 39,  9, 38, 33, 31,
       32, 11, 10,  5, 28,  4,  1, 13,  2, 17,  3, 34,  6, 14, 37, 35, 36,
       21, 26, 27, 25,  7, 15], dtype=int64)

In [344]:
base_df_linear2 = base_df_linear2.drop(base_df_linear2.loc[:, 'Wilderness_Area_1':'Cover_Type'].columns, axis = 1)

In [345]:
base_df_linear2.columns

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Hydrology',
       'Vertical_Distance_To_Hydrology', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points'],
      dtype='object')

In [346]:
base_df_linear2_1 = pd.DataFrame.copy(base_df_linear2)
base_df_linear2_2 = pd.DataFrame.copy(base_df_linear2)
base_df_linear2_3 = pd.DataFrame.copy(base_df_linear2)
Wilderness_Area_Bruto = pd.DataFrame.copy(data.iloc[0:,10:14])
Soil_Type_Bruto = pd.DataFrame.copy(data.iloc[0:,14:54])
Cover_Type = pd.DataFrame.copy(data.iloc[0:,54])

In [347]:
Wilderness_Area_Bruto.columns

Index(['Wilderness_Area_1', 'Wilderness_Area_2', 'Wilderness_Area_3',
       'Wilderness_Area_4'],
      dtype='object')

In [348]:
Soil_Type_Bruto.columns

Index(['Soil_Type_1', 'Soil_Type_2', 'Soil_Type_3', 'Soil_Type_4',
       'Soil_Type_5', 'Soil_Type_6', 'Soil_Type_7', 'Soil_Type_8',
       'Soil_Type_9', 'Soil_Type_10', 'Soil_Type_11', 'Soil_Type_12',
       'Soil_Type_13', 'Soil_Type_14', 'Soil_Type_15', 'Soil_Type_16',
       'Soil_Type_17', 'Soil_Type_18', 'Soil_Type_19', 'Soil_Type_20',
       'Soil_Type_21', 'Soil_Type_22', 'Soil_Type_23', 'Soil_Type_24',
       'Soil_Type_25', 'Soil_Type_26', 'Soil_Type_27', 'Soil_Type_28',
       'Soil_Type_29', 'Soil_Type_30', 'Soil_Type_31', 'Soil_Type_32',
       'Soil_Type_33', 'Soil_Type_34', 'Soil_Type_35', 'Soil_Type_36',
       'Soil_Type_37', 'Soil_Type_38', 'Soil_Type_39', 'Soil_Type_40'],
      dtype='object')

In [349]:
Cover_Type

0         5
1         5
2         2
3         2
4         5
         ..
581007    3
581008    3
581009    3
581010    3
581011    3
Name: Cover_Type, Length: 581012, dtype: int64

In [350]:
base_df_linear2_1 = pd.concat([base_df_linear2_1, Wilderness_Area, Soil_Type, Cover_Type], axis=1)
base_df_linear2_2 = pd.concat([base_df_linear2_2, Wilderness_Area_Bruto, Soil_Type, Cover_Type], axis=1)
base_df_linear2_3 = pd.concat([base_df_linear2_3, Wilderness_Area, Soil_Type_Bruto, Cover_Type], axis=1)

In [351]:
print(base_df_linear2_1.columns)
print(base_df_linear2_1.shape)

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Hydrology',
       'Vertical_Distance_To_Hydrology', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area', 'Soil_Type',
       'Cover_Type'],
      dtype='object')
(581012, 13)


In [352]:
print(base_df_linear2_2.columns)
print(base_df_linear2_2.shape)

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Hydrology',
       'Vertical_Distance_To_Hydrology', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area_1',
       'Wilderness_Area_2', 'Wilderness_Area_3', 'Wilderness_Area_4',
       'Soil_Type', 'Cover_Type'],
      dtype='object')
(581012, 16)


In [353]:
print(base_df_linear2_3.columns)
print(base_df_linear2_3.shape)

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Hydrology',
       'Vertical_Distance_To_Hydrology', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area', 'Soil_Type_1',
       'Soil_Type_2', 'Soil_Type_3', 'Soil_Type_4', 'Soil_Type_5',
       'Soil_Type_6', 'Soil_Type_7', 'Soil_Type_8', 'Soil_Type_9',
       'Soil_Type_10', 'Soil_Type_11', 'Soil_Type_12', 'Soil_Type_13',
       'Soil_Type_14', 'Soil_Type_15', 'Soil_Type_16', 'Soil_Type_17',
       'Soil_Type_18', 'Soil_Type_19', 'Soil_Type_20', 'Soil_Type_21',
       'Soil_Type_22', 'Soil_Type_23', 'Soil_Type_24', 'Soil_Type_25',
       'Soil_Type_26', 'Soil_Type_27', 'Soil_Type_28', 'Soil_Type_29',
       'Soil_Type_30', 'Soil_Type_31', 'Soil_Type_32', 'Soil_Type_33',
       'Soil_Type_34', 'Soil_Type_35', 'Soil_Type_36', 'Soil_Type_37',
       'Soil_Type_38', 'Soil_Type_39', 'Soil_Type_40', 'Cover_Type'],
      dtyp

In [354]:
X = base_df_linear2_1.iloc[0:,0:12]
y = base_df_linear2_1.iloc[0:,12]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.125356046125


In [355]:
X = base_df_linear2_2.iloc[0:,0:15]
y = base_df_linear2_2.iloc[0:,15]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.142908736426


In [356]:
X = base_df_linear2_3.iloc[0:,0:51]
y = base_df_linear2_3.iloc[0:,51]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323335501331


O melhor resultado segue o da base line, vou tentar normalizar os dados representados em metros

In [357]:
norm = base_df_linear2_1.loc[0:,['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]

In [358]:
p=MaxAbsScaler()
p.fit(norm)
MaxAbsScaler(copy=True)
norm = p.transform(norm)

In [359]:
norm = pd.DataFrame(norm, columns=['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points'])

In [360]:
norm

Elevation  Horizontal_Distance_To_Hydrology  \
0        0.672888                          0.184681   
1        0.671332                          0.151754   
2        0.726801                          0.191840   
3        0.721877                          0.173228   
4        0.672628                          0.109520   
...           ...                               ...   
581007   0.621047                          0.060845   
581008   0.619751                          0.047960   
581009   0.618455                          0.042949   
581010   0.617937                          0.042949   
581011   0.617678                          0.042949   

        Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                             0.000000                         0.071659   
1                            -0.009983                         0.054798   
2                             0.108153                         0.446817   
3                             0.196339                         0.434172   
4                            -0.001664                         0.054939   
...                                ...                              ...   
581007                        0.028286                         0.015175   
581008                        0.019967                         0.013348   
581009                        0.011647                         0.012646   
581010                        0.008319                         0.012646   
581011                        0.006656                         0.009414   

        Horizontal_Distance_To_Fire_Points  
0                                 0.875366  
1                                 0.867838  
2                                 0.853339  
3                                 0.865886  
4                                 0.860449  
...                                    ...  
581007                            0.116688  
581008                            0.117803  
581009                            0.119058  
581010                            0.120452  
581011                            0.121985  

[581012 rows x 5 columns]

In [361]:
base_df_linear2_1[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']] = norm[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]
base_df_linear2_2[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']] = norm[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]
base_df_linear2_3[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']] = norm[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]
base_df_linear2_4 = pd.DataFrame.copy(data)
base_df_linear2_4[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']] = norm[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]

In [362]:
X = base_df_linear2_1.iloc[0:,0:12]
y = base_df_linear2_1.iloc[0:,12]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.125356046125


In [363]:
X = base_df_linear2_2.iloc[0:,0:15]
y = base_df_linear2_2.iloc[0:,15]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.142908736426


In [364]:
X = base_df_linear2_3.iloc[0:,0:51]
y = base_df_linear2_3.iloc[0:,51]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323335501331


In [365]:
X = base_df_linear2_4.iloc[0:,0:54]
y = base_df_linear2_4.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323434740334


In [366]:
base_df_linear2_5 = pd.DataFrame.copy(data)
X = base_df_linear2_5.iloc[0:,0:54]
y = base_df_linear2_5.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression(normalize=True)  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323433827263


A base line, sem normalização e sem redução de dimensionalidade segue sendo a melhor proposta para esse algoritmo

#### 3.4) **(10%)** Tunning - Agora que temos um resultado promissor, vamos tentar melhorar o resultado alterando um ou mais hiper-parametro. Compare os resultados.

In [367]:
base_df_linear_f = pd.DataFrame.copy(data)

In [368]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323434740334


In [369]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.316501875774


In [370]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.320621820269


In [371]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.319947759201


In [372]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=4)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.32041652249


In [373]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=5)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.318493412754


In [374]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.326957277016


In [375]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.323434740334


In [376]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.320378851477


In [377]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.319658209109


In [378]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.319323790339


In [379]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.326957277016


In [380]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.09, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.3273851719


In [381]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.08, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.327688573151


In [382]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.07, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.328605980294


In [383]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.06, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.329690376235


In [384]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.05, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.329551902566


In [385]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.04, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.32952472058


In [386]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.03, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.33519308508


In [387]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.02, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.342112334937


In [388]:
X = base_df_linear_f.iloc[0:,0:54]
y = base_df_linear_f.iloc[0:,54]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.01, random_state=0)
regressor = LinearRegression()  
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Score:',sklearn.metrics.r2_score(y_test,y_pred))

Score: 0.334219389637


O melhot resultado deste algoritimo foi um Score de 0.342112334937 com test_size=0.02, random_state=0

#### 3.5) Qual método escolhido?

Arvore de decisão 

#### 3.6) **(10%)** Baseline - Implemente seu classificador da forma mais simples possível para esse ser seu baseline

In [389]:
base_df_tree = pd.DataFrame(data)

In [390]:
model = tree.DecisionTreeClassifier(random_state=0)

In [391]:
t1 = time.time()
X = base_df_tree.iloc[0:,0:54]
y = base_df_tree.iloc[0:,54]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.592135119979
14.190329551696777


#### 3.7) **(20%)** Versão 1 - O que podemos fazer para melhorar nosso baseline? Aplique técnicas como redução de dimensionalidade, normalização ou outras. Compare os resultados.

In [392]:
base_df_tree2 = base_df_tree.drop(base_df_tree.loc[:, 'Wilderness_Area_1':'Cover_Type'].columns, axis = 1)

In [393]:
base_df_tree2_1 = pd.DataFrame.copy(base_df_tree2)
base_df_tree2_2 = pd.DataFrame.copy(base_df_tree2)
base_df_tree2_3 = pd.DataFrame.copy(base_df_tree2)
Wilderness_Area_Bruto = pd.DataFrame.copy(data.iloc[0:,10:14])
Soil_Type_Bruto = pd.DataFrame.copy(data.iloc[0:,14:54])
Cover_Type = pd.DataFrame.copy(data.iloc[0:,54])

In [394]:
base_df_tree2_1 = pd.concat([base_df_tree2, Wilderness_Area, Soil_Type, Cover_Type], axis=1)
base_df_tree2_2 = pd.concat([base_df_tree2, Wilderness_Area_Bruto, Soil_Type, Cover_Type], axis=1)
base_df_tree2_3 = pd.concat([base_df_tree2, Wilderness_Area, Soil_Type_Bruto, Cover_Type], axis=1)

In [395]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0)
X = base_df_tree2_1.iloc[0:,0:12]
y = base_df_tree2_1.iloc[0:,12]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.585383165831
9.429008960723877


In [396]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0)
X = base_df_tree2_2.iloc[0:,0:15]
y = base_df_tree2_2.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.593797862028
8.675210952758789


In [397]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0)
X = base_df_tree2_3.iloc[0:,0:51]
y = base_df_tree2_3.iloc[0:,51]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.587771896592
12.143720865249634


In [398]:
# modelo com o melhor resultado ate o momento foi o da base base_df_tree2_2, vamos tentar mais algumas modificações.
base_df_tree2_2.columns

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Hydrology',
       'Vertical_Distance_To_Hydrology', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area_1',
       'Wilderness_Area_2', 'Wilderness_Area_3', 'Wilderness_Area_4',
       'Soil_Type', 'Cover_Type'],
      dtype='object')

In [399]:
for i in base_df_tree2_2.index:
    base_df_tree2_2.at[i, 'Elevation'] = round(((base_df_tree2_2.at[i, 'Elevation']-1859)/200),0)
    base_df_tree2_2.at[i, 'Horizontal_Distance_To_Hydrology'] = round(((base_df_tree2_2.at[i, 'Horizontal_Distance_To_Hydrology']-0)/140),0)
    base_df_tree2_2.at[i, 'Vertical_Distance_To_Hydrology'] = round(((base_df_tree2_2.at[i, 'Vertical_Distance_To_Hydrology']--173)/77),0)
    base_df_tree2_2.at[i, 'Horizontal_Distance_To_Roadways'] = round(((base_df_tree2_2.at[i, 'Horizontal_Distance_To_Roadways']-0)/712),0)
    base_df_tree2_2.at[i, 'Horizontal_Distance_To_Fire_Points'] = round(((base_df_tree2_2.at[i, 'Horizontal_Distance_To_Fire_Points']-0)/717),0)

In [400]:
pd.unique(base_df_tree2_2[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']].values.ravel('K'))
#Todas as colunas de informações em metros foram categorizadas em 10 grupos

array([ 4,  5,  3,  6,  7,  8,  1,  2,  0,  9, 10], dtype=int64)

In [401]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0)
X = base_df_tree2_2.iloc[0:,0:15]
y = base_df_tree2_2.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.565492171815
5.977886199951172


In [402]:
#Não deu muito certo, vamos voltar atraz e tentar a normalização
base_df_tree2_2 = pd.DataFrame.copy(base_df_tree2)
base_df_tree2_2 = pd.concat([base_df_tree2, Wilderness_Area_Bruto, Soil_Type, Cover_Type], axis=1)
base_df_tree2_2

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0            2596      51      3                               258   
1            2590      56      2                               212   
2            2804     139      9                               268   
3            2785     155     18                               242   
4            2595      45      2                               153   
...           ...     ...    ...                               ...   
581007       2396     153     20                                85   
581008       2391     152     19                                67   
581009       2386     159     17                                60   
581010       2384     170     15                                60   
581011       2383     165     13                                60   

        Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                                    0                              510   
1                                   -6                              390   
2                                   65                             3180   
3                                  118                             3090   
4                                   -1                              391   
...                                ...                              ...   
581007                              17                              108   
581008                              12                               95   
581009                               7                               90   
581010                               5                               90   
581011                               4                               67   

        Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                 221             232            148   
1                 220             235            151   
2                 234             238            135   
3                 238             238            122   
4                 220             234            150   
...               ...             ...            ...   
581007            240             237            118   
581008            240             237            119   
581009            236             241            130   
581010            230             245            143   
581011            231             244            141   

        Horizontal_Distance_To_Fire_Points  Wilderness_Area_1  \
0                                     6279                  1   
1                                     6225                  1   
2                                     6121                  1   
3                                     6211                  1   
4                                     6172                  1   
...                                    ...                ...   
581007                                 837                  0   
581008                                 845                  0   
581009                                 854                  0   
581010                                 864                  0   
581011                                 875                  0   

        Wilderness_Area_2  Wilderness_Area_3  Wilderness_Area_4  Soil_Type  \
0                       0                  0                  0         29   
1                       0                  0                  0         29   
2                       0                  0                  0         12   
3                       0                  0                  0         30   
4                       0                  0                  0         29   
...                   ...                ...                ...        ...   
581007                  0                  1                  0          2   
581008                  0                  1                  0          2   
581009                  0                  1                  0          2   
581010                  0       

In [403]:
norm = base_df_tree2.loc[0:,['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]
p=MaxAbsScaler()
p.fit(norm)
MaxAbsScaler(copy=True)
norm = p.transform(norm)
norm = pd.DataFrame(norm, columns=['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points'])
norm

Elevation  Horizontal_Distance_To_Hydrology  \
0        0.672888                          0.184681   
1        0.671332                          0.151754   
2        0.726801                          0.191840   
3        0.721877                          0.173228   
4        0.672628                          0.109520   
...           ...                               ...   
581007   0.621047                          0.060845   
581008   0.619751                          0.047960   
581009   0.618455                          0.042949   
581010   0.617937                          0.042949   
581011   0.617678                          0.042949   

        Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                             0.000000                         0.071659   
1                            -0.009983                         0.054798   
2                             0.108153                         0.446817   
3                             0.196339                         0.434172   
4                            -0.001664                         0.054939   
...                                ...                              ...   
581007                        0.028286                         0.015175   
581008                        0.019967                         0.013348   
581009                        0.011647                         0.012646   
581010                        0.008319                         0.012646   
581011                        0.006656                         0.009414   

        Horizontal_Distance_To_Fire_Points  
0                                 0.875366  
1                                 0.867838  
2                                 0.853339  
3                                 0.865886  
4                                 0.860449  
...                                    ...  
581007                            0.116688  
581008                            0.117803  
581009                            0.119058  
581010                            0.120452  
581011                            0.121985  

[581012 rows x 5 columns]

In [404]:
base_df_tree2_2[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']] = norm[['Elevation','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology','Horizontal_Distance_To_Roadways','Horizontal_Distance_To_Fire_Points']]

In [405]:
base_df_tree2_2

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0        0.672888      51      3                          0.184681   
1        0.671332      56      2                          0.151754   
2        0.726801     139      9                          0.191840   
3        0.721877     155     18                          0.173228   
4        0.672628      45      2                          0.109520   
...           ...     ...    ...                               ...   
581007   0.621047     153     20                          0.060845   
581008   0.619751     152     19                          0.047960   
581009   0.618455     159     17                          0.042949   
581010   0.617937     170     15                          0.042949   
581011   0.617678     165     13                          0.042949   

        Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                             0.000000                         0.071659   
1                            -0.009983                         0.054798   
2                             0.108153                         0.446817   
3                             0.196339                         0.434172   
4                            -0.001664                         0.054939   
...                                ...                              ...   
581007                        0.028286                         0.015175   
581008                        0.019967                         0.013348   
581009                        0.011647                         0.012646   
581010                        0.008319                         0.012646   
581011                        0.006656                         0.009414   

        Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                 221             232            148   
1                 220             235            151   
2                 234             238            135   
3                 238             238            122   
4                 220             234            150   
...               ...             ...            ...   
581007            240             237            118   
581008            240             237            119   
581009            236             241            130   
581010            230             245            143   
581011            231             244            141   

        Horizontal_Distance_To_Fire_Points  Wilderness_Area_1  \
0                                 0.875366                  1   
1                                 0.867838                  1   
2                                 0.853339                  1   
3                                 0.865886                  1   
4                                 0.860449                  1   
...                                    ...                ...   
581007                            0.116688                  0   
581008                            0.117803                  0   
581009                            0.119058                  0   
581010                            0.120452                  0   
581011                            0.121985                  0   

        Wilderness_Area_2  Wilderness_Area_3  Wilderness_Area_4  Soil_Type  \
0                       0                  0                  0         29   
1                       0                  0                  0         29   
2                       0                  0                  0         12   
3                       0                  0                  0         30   
4                       0                  0                  0         29   
...                   ...                ...                ...        ...   
581007                  0                  1                  0          2   
581008                  0                  1                  0          2   
581009                  0                  1                  0          2   
581010                  0       

In [406]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0)
X = base_df_tree2_2.iloc[0:,0:15]
y = base_df_tree2_2.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.593794420042
8.871583461761475


In [407]:
#O melhor resultado final foi:
base_df_tree2_2 = pd.DataFrame.copy(base_df_tree2)
base_df_tree2_2 = pd.concat([base_df_tree2, Wilderness_Area_Bruto, Soil_Type, Cover_Type], axis=1)
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0)
X = base_df_tree2_2.iloc[0:,0:15]
y = base_df_tree2_2.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.593797862028
8.967090368270874


#### 3.8) **(10%)** Tunning - Agora que temos um resultado promissor, vamos tentar melhorar o resultado alterando um ou mais hiper-parametro. Compare os resultados.

In [408]:
base_df_tree_f = pd.DataFrame.copy(base_df_tree2)
base_df_tree_f = pd.concat([base_df_tree2, Wilderness_Area_Bruto, Soil_Type, Cover_Type], axis=1)

In [409]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='gini',splitter='best')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.593797862028
8.520536422729492


In [410]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='gini',splitter='random')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.580663942949
3.359203577041626


In [411]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596802822412
8.831647872924805


In [412]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='random')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.589660174438
3.235978364944458


Os parametro criterion='entropy',splitter='best' se sariram melhor, vamos fazer mais algumas alterações partindo deles

In [413]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
8.858641624450684


In [414]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=4)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.594873375024
8.828983545303345


melhor config ate o momento criterion='entropy',splitter='best',min_samples_split=3

In [415]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.59161007239
8.829383611679077


In [416]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=2,min_samples_leaf=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.59161007239
8.807221412658691


In [417]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.583902705273
8.858686208724976


In [418]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=4)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.579493105434
8.883024215698242


In [419]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=0.5)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.487599223769
1.1642365455627441


In [420]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
9.28481388092041


melhor config ate o momento criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1

In [421]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=100)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
9.315679788589478


In [422]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=50)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
9.432775020599365


In [423]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=30)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596823450182
9.25381588935852


In [424]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
8.930741786956787


melhor config ate o momento:
criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None

In [425]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None,max_features='auto')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.569612397049
3.0676889419555664


In [426]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None,max_features='sqrt')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.569612397049
3.173055648803711


In [427]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None,max_features='log2')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.569612397049
3.2300777435302734


In [428]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None,max_features=None)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
9.926758289337158


melhor config ate o momento:
criterion='entropy',splitter='best',min_samples_split=3,min_samples_leaf=1,max_depth=None,max_features=None

melhor config ate o momento:
criterion='entropy',min_samples_split=3

In [429]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',min_samples_split=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596802822412
9.177619934082031


In [430]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=0,criterion='entropy',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597606586475
9.804231882095337


In [431]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=1,criterion='entropy',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.595721925301
9.606388330459595


In [432]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=2,criterion='entropy',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596498159879
9.278300046920776


In [433]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=3,criterion='entropy',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596339828376
9.158053159713745


In [434]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=4,criterion='entropy',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596618599208
9.073559999465942


In [435]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=5,criterion='entropy',min_samples_split=3)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597544608699
9.53110408782959


In [436]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=1,criterion='entropy',min_samples_split=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.59779077742
9.20990777015686


In [437]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=2,criterion='entropy',min_samples_split=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597530840166
10.41986870765686


In [438]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=3,criterion='entropy',min_samples_split=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597832023005
9.804798603057861


In [439]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=4,criterion='entropy',min_samples_split=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596754585474
9.658229112625122


In [440]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=5,criterion='entropy',min_samples_split=2)
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.596470680074
9.66217041015625


In [441]:
t1 = time.time()
model = tree.DecisionTreeClassifier(random_state=3,criterion='entropy')
X = base_df_tree_f.iloc[0:,0:15]
y = base_df_tree_f.iloc[0:,15]
print(cross_val_score(model, X, y, cv=10,n_jobs=-2).mean())
t2 = time.time()
print(t2-t1)

0.597832023005
9.272418975830078


melhor config final:
random_state=3,criterion='entropy'

### 5) **(10%)** Conclusões

*Compare seus resultados. Imaginando que sua solução fosse para produção, qual deles você escolheria? Por que? Quais os riscos você enxerga? O que recomendaria de próximos passos para melhorar os resultados?*

Quanto aos dois algoritmos, o primeiro performou muito mal nas previsões e mesmo com um treino de 99% da base não passou dos 35% de acerto, já o segundo se saiu bem melhor. Em um cenário de produção o melhor resultado a ser aplicado, até o momento, seria os DecisionTreeClassifier com parâmetros: random_state=3,criterion='entropy'. Eu escolheria este modelo pois, ele foi o que apresentou a maior precisão em CV 59,78%, ainda é um valor que considero baixo para um ambiente de produção devido ao acerto ser pouco melhor que a escolha randômica. Os próximos passos para o aprimoramento dos resultados seria o teste de outros algoritmos, em busca de uma que melhor se adeque a situação.